File: packages/sdk-typescript/src/span.ts
import { randomUUID } from 'node:crypto';

export interface SpanData {
  id: string;
  trace_id: string;
  parent_span_id: string | null;
  name: string;
  type: string;
  input: string | null;
  output: string | null;
  started_at: string;
  ended_at: string | null;
  error: string | null;
  /** Optional session grouping. Populated from the current session() context
   *  or via trace(fn, { sessionId }). Default null preserves prior behavior. */
  session_id: string | null;
  /** Optional discriminator for spans that originate from a typed
   *  integration. Currently used for Claude extended-thinking — the
   *  integration emits `claude_call` parents with `thinking` and
   *  `response` children. Default null = ordinary span. */
  span_subtype?: 'thinking' | 'response' | string | null;
  /** Estimated tokens spent inside a thinking block. Only set on
   *  thinking spans. Anthropic doesn't break thinking out of
   *  output_tokens — we estimate from text length. */
  thinking_tokens?: number | null;
  /** LLM model identifier (e.g. "claude-opus-4-20250514"). */
  model?: string | null;
  provider?: string | null;
  tokens_input?: number | null;
  tokens_output?: number | null;
  cost_usd?: number | null;
}

function serialize(value: unknown, maxSize: number): string | null {
  if (value === null || value === undefined) return null;
  let s: string;
  try {
    s = JSON.stringify(value);
    if (s === undefined) s = String(value);
  } catch {
    try {
      s = String(value);
    } catch {
      s = '<unserializable>';
    }
  }
  return s.length > maxSize ? s.slice(0, maxSize) : s;
}

export class Span {
  readonly id: string;
  readonly trace_id: string;
  readonly parent_span_id: string | null;
  readonly name: string;
  readonly type: string;
  readonly started_at: string;
  input: string | null = null;
  output: string | null = null;
  ended_at: string | null = null;
  error: string | null = null;
  session_id: string | null = null;
  span_subtype: 'thinking' | 'response' | string | null = null;
  thinking_tokens: number | null = null;
  model: string | null = null;
  provider: string | null = null;
  tokens_input: number | null = null;
  tokens_output: number | null = null;
  cost_usd: number | null = null;

  private constructor(
    id: string,
    trace_id: string,
    parent_span_id: string | null,
    name: string,
    type: string,
  ) {
    this.id = id;
    this.trace_id = trace_id;
    this.parent_span_id = parent_span_id;
    this.name = name;
    this.type = type;
    this.started_at = new Date().toISOString();
  }

  static create(name: string, type = 'custom', parent?: Span | null): Span {
    const id = randomUUID();
    if (parent) {
      return new Span(id, parent.trace_id, parent.id, name, type);
    }
    return new Span(id, id, null, name, type);
  }

  setInput(value: unknown, maxSize = 10_240): void {
    this.input = serialize(value, maxSize);
  }

  setOutput(value: unknown, maxSize = 10_240): void {
    this.output = serialize(value, maxSize);
  }

  setError(err: unknown): void {
    if (err instanceof Error) {
      this.error = `${err.constructor.name}: ${err.message}`;
    } else {
      this.error = String(err);
    }
  }

  end(): void {
    if (this.ended_at === null) {
      this.ended_at = new Date().toISOString();
    }
  }

  toJSON(): SpanData {
    return {
      id: this.id,
      trace_id: this.trace_id,
      parent_span_id: this.parent_span_id,
      name: this.name,
      type: this.type,
      input: this.input,
      output: this.output,
      started_at: this.started_at,
      ended_at: this.ended_at,
      error: this.error,
      session_id: this.session_id,
      span_subtype: this.span_subtype,
      thinking_tokens: this.thinking_tokens,
      model: this.model,
      provider: this.provider,
      tokens_input: this.tokens_input,
      tokens_output: this.tokens_output,
      cost_usd: this.cost_usd,
    };
  }
}

File: packages/sdk-python/korveo/span.py
import json
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Optional


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _to_json(value: Any, max_size: int) -> Optional[str]:
    if value is None:
        return None
    try:
        s = json.dumps(value, default=str, ensure_ascii=False)
    except (TypeError, ValueError):
        try:
            s = json.dumps(str(value), ensure_ascii=False)
        except Exception:
            s = json.dumps("<unserializable>")
    if len(s) > max_size:
        s = s[:max_size]
    return s


@dataclass
class Span:
    id: str
    trace_id: str
    parent_span_id: Optional[str]
    name: str
    type: str = "custom"
    input: Optional[str] = None
    output: Optional[str] = None
    started_at: str = field(default_factory=_now_iso)
    ended_at: Optional[str] = None
    error: Optional[str] = None
    # Optional session grouping — populated from korveo.session() context
    # or @korveo.trace(session_id=...). Default None preserves existing
    # behavior for callers that don't use sessions.
    session_id: Optional[str] = None

    @classmethod
    def create(
        cls,
        name: str,
        type: str = "custom",
        parent: Optional["Span"] = None,
    ) -> "Span":
        new_id = str(uuid.uuid4())
        if parent is None:
            trace_id = new_id
            parent_span_id = None
        else:
            trace_id = parent.trace_id
            parent_span_id = parent.id
        return cls(
            id=new_id,
            trace_id=trace_id,
            parent_span_id=parent_span_id,
            name=name,
            type=type,
        )

    def end(self) -> None:
        if self.ended_at is None:
            self.ended_at = _now_iso()

    def set_input(self, value: Any, max_size: int = 10_240) -> None:
        self.input = _to_json(value, max_size)

    def set_output(self, value: Any, max_size: int = 10_240) -> None:
        self.output = _to_json(value, max_size)

    def set_error(self, exc: BaseException) -> None:
        self.error = f"{type(exc).__name__}: {exc}"

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "trace_id": self.trace_id,
            "parent_span_id": self.parent_span_id,
            "name": self.name,
            "type": self.type,
            "input": self.input,
            "output": self.output,
            "started_at": self.started_at,
            "ended_at": self.ended_at,
            "error": self.error,
            "session_id": self.session_id,
        }


class SpanContext:
    """Context manager for a span. Used as `with korveo.span("name"):`."""

    def __init__(self, sdk, name: str, type: str = "custom"):
        self._sdk = sdk
        self._name = name
        self._type = type
        self._span: Optional[Span] = None
        self._token = None

    def __enter__(self) -> Span:
        from .context import get_current_span, set_current_span

        parent = get_current_span()
        self._span = Span.create(self._name, self._type, parent=parent)
        self._token = set_current_span(self._span)
        return self._span

    def __exit__(self, exc_type, exc_val, exc_tb) -> bool:
        from .context import reset_current_span

        assert self._span is not None
        if exc_val is not None:
            self._span.set_error(exc_val)
        self._span.end()
        self._sdk.submit(self._span)
        if self._token is not None:
            reset_current_span(self._token)
        return False
